### PR TITLE
gfx-replay recording for C++ viewer

### PR DESCRIPTION
**Pull request by eundersander**
## Motivation and Context

This adds gfx-replay recording to the C++ viewer. Specify an output filepath via command line. A keyframe is saved after every physics step. Press R to write all keyframes to the file. Every time you press R, you'll overwrite the old replay file with new keyframes.

## How Has This Been Tested

Ad hoc testing in the viewer.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/esp/gfx/replay/Recorder.h [hunk 1]
```diff
@@ -90,13 +90,13 @@ class Recorder {
                                   const Magnum::Quaternion& rotation);
 
   /**
-   * @brief write saved keyframes to file. Not implemented yet.
+   * @brief write saved keyframes to file.
```
Comment (eundersander) [inline src/esp/gfx/replay/Recorder.h:93]: Unrelated fix: these comments were outdated.

File: src/utils/viewer/viewer.cpp [hunk 2]
```diff
@@ -343,6 +346,9 @@ Viewer::Viewer(const Arguments& arguments)
                "Stage asset should be lit with Phong shading.")
       .addBooleanOption("debug-bullet")
       .setHelp("debug-bullet", "Render Bullet physics debug wireframes.")
+      .addOption("gfx-replay-record-filepath")
+      .setHelp("gfx-replay-record-filepath",
+               "Enable replay recording with R key.")
```
Comment (aclegg3) [inline src/utils/viewer/viewer.cpp:351]: Add this key to the helpText_ string at line 146 also.

File: src/utils/viewer/viewer.cpp [hunk 3]
```diff
@@ -388,12 +394,15 @@ Viewer::Viewer(const Arguments& arguments)
...
 
+  gfxReplayRecordFilepath_ = args.value("gfx-replay-record-filepath");
+
   // configure and intialize Simulator
   auto simConfig = esp::sim::SimulatorConfiguration();
   simConfig.activeSceneID = sceneFileName;
   simConfig.enablePhysics = useBullet;
   simConfig.frustumCulling = true;
   simConfig.requiresTextures = true;
+  simConfig.enableGfxReplaySave = !gfxReplayRecordFilepath_.empty();
```
Comment (aclegg3) [inline src/utils/viewer/viewer.cpp:405]: If this is false, `simulator_->getGfxReplayManager()->getRecorder()` returns nullptr?

File: src/utils/viewer/viewer.cpp [hunk 4]
```diff
@@ -388,12 +394,15 @@ Viewer::Viewer(const Arguments& arguments)
     debugBullet_ = true;
   }
 
+  gfxReplayRecordFilepath_ = args.value("gfx-replay-record-filepath");
```
Comment (aclegg3) [inline src/utils/viewer/viewer.cpp:397]: Could use a path join to support relative and global paths. Maybe this is already in the save function?

**Review by aclegg3 — APPROVED** [review]
Looks good. Cool feature.

File: src/utils/viewer/viewer.cpp (hunk 3 above)
Comment (eundersander) [reply src/utils/viewer/viewer.cpp:405]: yes (er, null shared_ptr)

File: src/utils/viewer/viewer.cpp (hunk 4 above)
Comment (eundersander) [reply src/utils/viewer/viewer.cpp:397]: I pass the filepath string directly to io::writeJsonToFile. With the code as is, you can do relative or absolute path like so:

./build/utils/viewer/viewer /my_stage.glb --gfx-replay-record-filepath ./relative/path/to/my_replay.json
./build/utils/viewer/viewer /my_stage.glb --gfx-replay-record-filepath /absolute/path/to/my_replay.json

The relative path is relative to the current working directory where you're running viewer.

Did you have something else in mind?

Comment (aclegg3) [reply src/utils/viewer/viewer.cpp:398]: That's fine. Just checking since we do this join explicitly for some other arguments (e.g. `"physics-config"` and `"object-dir"`).
